corehq/apps/users/management/commands/migrate_user_history_to_new_structure.py

```diff
@@ -0,0 +1,734 @@
+
+# location-name[location_id]
+# group-name[group_id]
+NAME_WITH_ID_REGEX = re.compile(r'(.+)\[(.+)]')
+
+
+class Command(BaseCommand):
+    help = "Migrate User History records to new structure"
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            '--save',
+            action='store_true',
+            dest='save',
+            default=False,
+            help="actually save records else just log",
+        )
+        parser.add_argument(
+            '--skip-assertions',
+            action='store_true',
+            dest='skip_assertions',
+            default=False,
+            help="to be used to skip assertions in change message creation, useful for manual updates",
+        )
+        parser.add_argument(
+            '--only-pending',
+            action='store_true',
+            dest='only_pending',
+            default=False,
+            help="to iterate only logs that are pending to be migrated"
+        )
+
+    def handle(self, **options):
+        save = options['save']
+        skip_assertions = options['skip_assertions']
+        only_pending = options['only_pending']
+        if save and skip_assertions:
+            raise CommandError("It is not recommended to have save with assertions skipped")
+        wb = Workbook()
+        ws = wb.active
+        ws.append([
+            'ID', 'Message', 'Change Messages', 'Details Changes', 'Changes', 'Changes equal?',
+            'Details Changed Via', 'Changed Via']
+        )
+
+        records = UserHistory.objects
+        if only_pending:
+            # filter on the new column that is expected to be blank for not migrated records only
+            records = records.filter(changed_via='')
+
+        for user_history in records.all():
+            try:
+                migrate(user_history, save=save, skip_assertions=skip_assertions)
+            except Exception as e:
+                logging.error(f"{user_history.pk}: {e}")
+            else:
+                ws.append(
+                    [user_history.pk,
+                     user_history.message,
+                     json.dumps(user_history.change_messages),
+                     json.dumps(user_history.details.get('changes')),
+                     json.dumps(user_history.changes),
+                     user_history.details.get('changes') == user_history.changes,  # should not be False
+                     user_history.details.get('changed_via'),
+                     user_history.changed_via
+                     ]
+                )
+        wb.save("migrate_user_history_to_new_structure.xlsx")
+
+
+def migrate(user_history, save=False, skip_assertions=False):
+    """
+    1. Copy over changed_via and changes to new columns
+    2. convert messages into new change messages format
+    3. migrate over to log entry to back up old columns before they are deleted
+    """
+    # ToDo: remove this if later. Just added to avoid issues during dry runs
```

Comment:
Just if block, yes. It would always be `True` after the [migration in the base PR](https://github.com/dimagi/commcare-hq/pull/30253/files#diff-3efade2c2ce1fb66eb147d588d776c09efd8bc0cb22db2ecdc6fc68ad07e8d79R16) is run